design/src/rfcs/rfc0020_service_builder.md

```diff
@@ -0,0 +1,719 @@
+let operation_1 = Operation1::from_service(/* some service */);
+
+OperationRegistryBuilder::default()
+    .operation0(operation_0)
+    .operation1(operation_1)
+    /* ... */
+```
+
+### Approach C: Operations as Middleware Constructors
+
+While [Attempt B](#approach-b-operations-as-middleware) solves all three problems, it fails to adequately model the Smithy semantics. An operation cannot uniquely define a `tower::Service` without reference to a parent Smithy service - information concerning the serialization/deserialization, error modes are all inherited from the Smithy service an operation is used within. In this way, `Operation0` should not be a standalone middleware, but become middleware once accepted by the service builder.
+
+Any solution which provides an `{Operation}` structure and wishes it to be accepted by multiple service builders must deal with this problem. We currently build one library per service and hence have duplicate structures when [service closures](https://awslabs.github.io/smithy/1.0/spec/core/model.html?highlight=closure#service-closure) overlap. This means we wouldn't run into this problem today, but it would be a future obstruction if we wanted to reduce the amount of generated code.
+
+```rust
+use tower::layer::util::{Stack, Identity};
+use tower::util::{ServiceFn, service_fn};
+
+// This takes the same form as `Operation0` defined in the previous attempt. The difference being that this is now
+// private.
+struct Service0Operation0<S> {
+    inner: S
+}
+
+impl<S> Service<http::Request> for ServiceOperation0<S>
+where
+    S: Service<Operation0Input, Response = Operation0Output, Error = Infallible>
+{
+    /* Same as above */
+}
+
+pub struct Operation0<S, L> {
+    inner: S,
+    layer: L
+}
+
+impl<S> Operation0<S, Identity> {
+    pub fn from_service(inner: S) -> Self {
+        Self {
+            inner,
+            layer: Identity
+        }
+    }
+}
+
+impl<F> Operation0<ServiceFn<F>, Identity> {
+    pub fn from_handler(inner: F) -> Self {
+        Operation0::from_service(service_fn(inner))
+    }
+}
+
+impl<S, L> Operation0<S, L> {
+    pub fn layer<NewL>(self, layer: L) -> Operation0<S, Stack<L, NewL>> {
+        Operation0 {
+            inner: self.inner,
+            layer: Stack::new(self.layer, layer)
+        }
+    }
+
+    pub fn logging(self, /* args */) -> Operation0<S, Stack<L, LoggingLayer>> {
+        Operation0 {
+            inner: self.inner,
+            layer: Stack::new(self.layer, LoggingLayer::new(/* args */))
+        }
+    }
+
+    pub fn auth(self, /* args */) -> Operation0<S, Stack<L, AuthLayer>> {
+        Operation0 {
+            inner: self.inner,
+            layer: Stack::new(self.layer, /* Construct auth middleware */)
+        }
+
+    }
+}
+
+impl<Op1, Op2> OperationRegistryBuilder<Op1, Op2> {
+    pub fn operation0<S, L>(self, operation: Operation0<S, L>) -> OperationRegistryBuilder<<L as Layer<Service0Operation0<S>>::Service, Op2>
```

Comment:
I like approach C.

Perhaps I'm asking too much, but wouldn't it be nice if we could make `operation0` take in _both_ an `Operation0` struct, and a function/closure just as it does now? So `registry.operation0(fun)` is just syntax sugar for `registry.operation0(Operation0::from_handler(fun))`. Maybe by making `operation0` take a generic param implementing a trait that we auto-implement in codegen for both `Operation0` and the correct function types (with and without state)? Essentially an `UberHandler`.

Reply:
I think this is probably definitely doable and desirable. We could have `operation0` take a `Into<Operation<S, L>>` and implement `From<F>` where `F: Fn(/* with or without state */)` perhaps.

One thing I'm struggling with is that there's a lot of moving parts in this section and it's hard to maintain a narrative without side stepping. I'll make sure to get this in though.

Reply:
On second thought I don't think it's possible. It's possible to `impl Handler` for `F: Fn(X) -> ...` and `F: Fn(X, Y) -> ...` because it's sugar for `Fn<(X, )>` and `Fn<(X, Y)>` - no struct can be both so we can't run into conflicts. This isn't the case for `F: Fn` and `S: Service<Req>` - structs could presumably be both.

Reply:
On second thought I'm not sure this is possible because a `S: Service<Input, Response = Output>` could also be an async `Input -> Output`, in which case there would be conflicting conversion implementations.